backupccl: break read and write metrics by bucket

andrewbaptist opened the pull request:
Previously the read and write metrics were aggregate across all
containers or buckets. In some situations it is useful to break the
statistics out to detemine how much data is going to each. This PR
allows breaking the data out for the three main clouds.

Epic: [CRDB-41142](https://cockroachlabs.atlassian.net/browse/CRDB-41142)

Release note: None

Commit:
backupccl: break read and write metrics by bucket

Previously the read and write metrics were aggregate across all
containers or buckets. In some situations it is useful to break the
statistics out to detemine how much data is going to each. This PR
allows breaking the data out for the three main clouds.

Epic: CRDB-41142

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129059)
<!-- Reviewable:end -->

Commit:
util: add connection metric tracking utility

Adds a utility to cidrLookup to created a DialContext that is tracked
based on the lookup. This works for any third-party libraries that
expose a way to set the DialContext.

Epic: none

Release note: None

jeffswenson commented on pkg/util/cidr/cidr.go:
```diff
@@ -326,3 +330,140 @@ func (c *Lookup) LookupIP(ip net.IP) string {
+}
+
+// NetMetrics are aggregate metrics around net.Conn mapped based on the CIDR lookup.
+type NetMetrics struct {
```
I think NetMetrics could use some unit tests.

jeffswenson reviewed (approved):
LGTM

andrewbaptist commented:
TFTR!

Jeff, I will add unit tests for `NetMetrics` in the followup PR for the changefeeds. I wanted to get this in to allow @adwittumuluri to continue testing and integration but I agree the unit testing is insufficient on this type.

bors r=jeffswenson
